<h2 align='center' style='color:blue'>Dropout Regularization In Deep Neural Network</h2>

This is a dataset that describes sonar chirp returns bouncing off different services. The 60 input variables are the strength of the returns at different angles. It is a **binary classification problem** that requires a model to differentiate rocks from metal cylinders.

Dataset information: https://archive.ics.uci.edu/ml/datasets/Connectionist+Bench+(Sonar,+Mines+vs.+Rocks)
Download it from here: https://archive.ics.uci.edu/ml/machine-learning-databases/undocumented/connectionist-bench/sonar/sonar.all-data

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import tensorflow as tf
from tensorflow import keras
from sklearn.metrics import confusion_matrix , classification_report
from sklearn.model_selection import train_test_split

import warnings
warnings.filterwarnings('ignore')

In [2]:
df = pd.read_csv("../Data/sonar_dataset.csv", header=None)
df.sample(5)

0       1       2       3       4       5       6       7       8   \
103  0.0162  0.0253  0.0262  0.0386  0.0645  0.0472  0.1056  0.1388  0.0598   
113  0.0283  0.0599  0.0656  0.0229  0.0839  0.1673  0.1154  0.1098  0.1370   
155  0.0211  0.0128  0.0015  0.0450  0.0711  0.1563  0.1518  0.1206  0.1666   
91   0.0253  0.0808  0.0507  0.0244  0.1724  0.3823  0.3729  0.3583  0.3429   
64   0.0071  0.0103  0.0135  0.0494  0.0253  0.0806  0.0701  0.0738  0.0117   

         9   ...      51      52      53      54      55      56      57  \
103  0.1334  ...  0.0071  0.0082  0.0232  0.0198  0.0074  0.0035  0.0100   
113  0.1767  ...  0.0147  0.0170  0.0158  0.0046  0.0073  0.0054  0.0033   
155  0.1345  ...  0.0117  0.0023  0.0047  0.0049  0.0031  0.0024  0.0039   
91   0.2197  ...  0.0178  0.0073  0.0079  0.0038  0.0116  0.0033  0.0039   
64   0.0898  ...  0.0043  0.0048  0.0076  0.0124  0.0105  0.0054  0.0032   

         58      59  60  
103  0.0048  0.0019   M  
113  0.0045  0.0079   M  
155  0.0051  0.0015   M  
91   0.0081  0.0053   R  
64   0.0073  0.0063   R  

[5 rows x 61 columns]

In [3]:
df.shape

(208, 61)

In [4]:
# check for nan values
df.isna().sum()

0     0
1     0
2     0
3     0
4     0
     ..
56    0
57    0
58    0
59    0
60    0
Length: 61, dtype: int64

In [5]:
df.columns

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
            34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
            51, 52, 53, 54, 55, 56, 57, 58, 59, 60],
           dtype='int64')

In [6]:
df[60].value_counts() # label is not skewed

M    111
R     97
Name: 60, dtype: int64

In [7]:
#Set the features for testing 
X = df.drop(60, axis=1)
y = df[60]
y.head()

0    R
1    R
2    R
3    R
4    R
Name: 60, dtype: object

In [8]:
#One hot encoding
y = pd.get_dummies(y, drop_first=True)
y.sample(5) # R --> 1 and M --> 0

R
47  1
92  1
15  1
94  1
10  1

In [9]:
y.value_counts()

R
0    111
1     97
dtype: int64

In [10]:
#test and train split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

In [11]:
X_test.shape

(52, 60)

In [12]:
y_test.shape

(52, 1)

In [13]:
X_train.shape

(156, 60)

## Model without Dropout Layer

In [46]:
model = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')
])

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [ ]:
model.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 1s 6ms/step - loss: 0.6988 - accuracy: 0.4227
Epoch 2/100
20/20 [==============================] - 0s 6ms/step - loss: 0.6692 - accuracy: 0.5651
Epoch 3/100
20/20 [==============================] - 0s 6ms/step - loss: 0.6347 - accuracy: 0.6703
Epoch 4/100
20/20 [==============================] - 0s 6ms/step - loss: 0.6286 - accuracy: 0.7186
Epoch 5/100
20/20 [==============================] - 0s 5ms/step - loss: 0.5875 - accuracy: 0.6934
Epoch 6/100
20/20 [==============================] - 0s 5ms/step - loss: 0.5521 - accuracy: 0.7557
Epoch 7/100
20/20 [==============================] - 0s 5ms/step - loss: 0.5185 - accuracy: 0.7403
Epoch 8/100
20/20 [==============================] - 0s 5ms/step - loss: 0.4786 - accuracy: 0.7855
Epoch 9/100
20/20 [==============================] - 0s 5ms/step - loss: 0.4882 - accuracy: 0.8073
Epoch 10/100
20/20 [==============================] - 0s 6ms/step - loss: 0.4596 - accuracy: 0.8229
Epoch 11/

20/20 [==============================] - 0s 5ms/step - loss: 0.0069 - accuracy: 1.0000
Epoch 84/100
20/20 [==============================] - 0s 6ms/step - loss: 0.0062 - accuracy: 1.0000
Epoch 85/100
20/20 [==============================] - 0s 5ms/step - loss: 0.0051 - accuracy: 1.0000
Epoch 86/100
20/20 [==============================] - 0s 5ms/step - loss: 0.0063 - accuracy: 1.0000
Epoch 87/100
20/20 [==============================] - 0s 5ms/step - loss: 0.0084 - accuracy: 1.0000
Epoch 88/100
20/20 [==============================] - 0s 5ms/step - loss: 0.0047 - accuracy: 1.0000
Epoch 89/100
20/20 [==============================] - 0s 6ms/step - loss: 0.0043 - accuracy: 1.0000
Epoch 90/100
20/20 [==============================] - 0s 6ms/step - loss: 0.0039 - accuracy: 1.0000
Epoch 91/100
20/20 [==============================] - 0s 6ms/step - loss: 0.0049 - accuracy: 1.0000
Epoch 92/100
20/20 [==============================] - 0s 5ms/step - loss: 0.0040 - accuracy: 1.0000
Epoch 93/100


In [45]:
model.evaluate(X_test, y_test)

2/2 [==============================] - 0s 4ms/step - loss: 1.2290 - accuracy: 0.7500


[1.2289974689483643, 0.75]

Training Accuracy >>> Test Accuracy

In [22]:
y_pred = model.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

[4.6401689e-07 9.9009949e-01 6.6855103e-01 4.5798617e-05 9.9999905e-01
 9.9415576e-01 4.3831265e-01 9.9999893e-01 4.6074114e-05 9.9999928e-01]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [23]:
y_test[:10]

R
186  0
155  0
165  0
200  0
58   1
34   1
151  0
18   1
202  0
62   1

In [24]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.71      0.89      0.79        27
           1       0.83      0.60      0.70        25

    accuracy                           0.75        52
   macro avg       0.77      0.74      0.74        52
weighted avg       0.77      0.75      0.74        52



### Model with Dropout Layer

In [41]:
modeld = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(240, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(120, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(1, activation='sigmoid')
])

modeld.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [ ]:
modeld.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 1s 6ms/step - loss: 0.7064 - accuracy: 0.4956
Epoch 2/100
20/20 [==============================] - 0s 6ms/step - loss: 0.7035 - accuracy: 0.4914
Epoch 3/100
20/20 [==============================] - 0s 6ms/step - loss: 0.6893 - accuracy: 0.5834
Epoch 4/100
20/20 [==============================] - 0s 6ms/step - loss: 0.6925 - accuracy: 0.5445
Epoch 5/100
20/20 [==============================] - 0s 6ms/step - loss: 0.6759 - accuracy: 0.6429
Epoch 6/100
20/20 [==============================] - 0s 6ms/step - loss: 0.7046 - accuracy: 0.5102
Epoch 7/100
20/20 [==============================] - 0s 6ms/step - loss: 0.6330 - accuracy: 0.6671
Epoch 8/100
20/20 [==============================] - 0s 7ms/step - loss: 0.6102 - accuracy: 0.6362
Epoch 9/100
20/20 [==============================] - 0s 6ms/step - loss: 0.6305 - accuracy: 0.6298
Epoch 10/100
20/20 [==============================] - 0s 6ms/step - loss: 0.5929 - accuracy: 0.7499
Epoch 11/

In [18]:
modeld.evaluate(X_test, y_test)

NameError: name 'modeld' is not defined

Training Accuracy is still good but Test Accuracy Improved

In [24]:
y_pred = modeld.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

[1.6179042e-06 7.8016031e-01 8.6281878e-01 4.0221727e-03 9.8407274e-01
 8.8602865e-01 7.1879250e-01 9.8856676e-01 7.9914415e-03 9.8803920e-01]
[0. 1. 1. 0. 1. 1. 1. 1. 0. 1.]


In [25]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.76      0.81      0.79        27
           1       0.78      0.72      0.75        25

    accuracy                           0.77        52
   macro avg       0.77      0.77      0.77        52
weighted avg       0.77      0.77      0.77        52



**You can see that by using dropout layer test accuracy increased from 0.77 to 0.81**